src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs

```diff
@@ -245,56 +263,70 @@ private static void RecordRequestStartEventLog(HttpContext httpContext)
         }
 
         [MethodImpl(MethodImplOptions.NoInlining)]
-        private Activity StartActivity(HttpContext httpContext, out bool hasDiagnosticListener)
+        private Activity? StartActivity(HttpContext httpContext, bool diagnosticListenerActivityCreationEnabled, out bool hasDiagnosticListener)
         {
-            var activity = new Activity(ActivityName);
             hasDiagnosticListener = false;
+            if (diagnosticListenerActivityCreationEnabled)
+            {
+                if (_diagnosticListener.IsEnabled(ActivityStartKey))
+                {
+                    hasDiagnosticListener = true;
+                }
+            }
+
+            // Short-circuit to avoid doing an expensive header lookup
+            if (!(hasDiagnosticListener || _activitySource.HasListeners()))
+            {
+                return null;
+            }
 
             var headers = httpContext.Request.Headers;
             if (!headers.TryGetValue(HeaderNames.TraceParent, out var requestId))
             {
                 headers.TryGetValue(HeaderNames.RequestId, out requestId);
             }
-
-            if (!StringValues.IsNullOrEmpty(requestId))
+            var activity = _activitySource.StartActivity(ActivityName, ActivityKind.Server, requestId);
```

Comment:
@tarekgh/@noahfalk, what's the expected format of the activity name when used with the ActivitySource?

I thought that considering the `ActivitySource` name is `Microsoft.AspNetCore.Hosting` the activity name should be `HttpRequestIn` and not `Microsoft.AspNetCore.Hosting.HttpRequestIn`

Reply:
You are correct @pakrym, the Activity names should be only the instrumented operation name (i.e. `HttpRequestIn` for this specific code). And ActivitySource name should reflect the company name, component name, and the whole instrumented area (i.e. `Microsoft.AspNetCore.Hosting`).

As this code existed for a while, do you think there will be any issue changing the Activity name?

Reply:
I think that's another area where reusing the same activity doesn't work well. We can't change existing activity names and ActivitySource-activity should have a different one.

Reply:
We spoke about this earlier today, the Activity name would have to be `Microsoft.AspNetCore.Hosting.HttpRequestIn` to preserve. Instrumentation libraries can set Activity.DisplayName if they would prefer something more concise.

Reply:
Should we say there is no guidance around `Activity.Name` then? If none of the Microsoft libraries would support it.

Reply:
The guidelines for Activity.Name still hold we only relax it for the sake of the compatibility. Any new written code with no compatibility concern should apply the guidelines.

Reply:
But if all the consuming code (OT) would forever have to check for legacy behavior. What's the point of the guideline?

Reply:
I am hoping the OTel will get rid of listening to the legacy behavior. My understanding is after this change, they are listening to the legacy for the payload and not the Activity itself. This may be the next step to look at that and try help OTel to get rid of this legacy listening.

Reply:
>I am hoping the OTel will get rid of listening to the legacy behavior

All ASP.NET Core ActivitySource activities would forever have legacy Activity.Name

Reply:
As I mentioned we relaxed the guidelines for the Activity.Name for the sake of compatibility. If anyone writing any new code with no compatibility concern, would better to follow the guidelines. For aspnet case, I suggest setting the display name.